components/email-mgt/org.wso2.carbon.email.mgt/src/main/java/org/wso2/carbon/email/mgt/NotificationTemplateManagerImpl.java

```diff
@@ -0,0 +1,720 @@
+                tenantDomain = getRootOrgTenantDomain(tenantDomain);
+            }
+            assertTemplateTypeExists(templateDisplayName, notificationChannel, tenantDomain);
+            List<NotificationTemplate> notificationTemplates =
+                    userDefinedTemplatePersistenceManager.listNotificationTemplates(templateDisplayName,
+                            notificationChannel, applicationUuid, tenantDomain);
+            if (notificationTemplates == null) {
+                notificationTemplates = new ArrayList<>();
+            }
+            return notificationTemplates;
+        } catch (OrganizationManagementException e) {
+            throw handleServerException(e.getMessage(), e);
+        }
+    }
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public NotificationTemplate getNotificationTemplate(String notificationChannel, String templateType, String locale,
+                                                         String tenantDomain)
+            throws NotificationTemplateManagerException {
+
+        return getNotificationTemplate(notificationChannel, templateType, locale, tenantDomain, null);
+    }
+
+    /**
+     * {@inheritDoc}
+     */
+    @Override
+    public NotificationTemplate getNotificationTemplate(String notificationChannel, String templateType, String locale,
+                                                         String tenantDomain, String applicationUuid)
+            throws NotificationTemplateManagerException {
+
+        try {
+            if (OrganizationManagementUtil.isOrganization(tenantDomain)) {
+                // To return the root organization's notification template.
+                tenantDomain = getRootOrgTenantDomain(tenantDomain);
+                // If it's application specific template is required, get the root organization's application.
+                if (StringUtils.isNotBlank(applicationUuid)) {
+                    applicationUuid = I18nMgtDataHolder.getInstance().getApplicationManagementService()
+                            .getMainAppId(applicationUuid);
+                }
+            }
+        } catch (OrganizationManagementException e) {
+            throw new NotificationTemplateManagerException(e.getMessage(), e);
+        } catch (IdentityApplicationManagementServerException e) {
+            String code = I18nEmailUtil.prependOperationScenarioToErrorCode(
+                    TemplateMgtConstants.ErrorMessages.ERROR_CODE_ERROR_RESOLVING_MAIN_APPLICATION.getCode(),
+                    TemplateMgtConstants.ErrorScenarios.NOTIFICATION_TEMPLATE_MANAGER);
+            String message = String.format(
+                    TemplateMgtConstants.ErrorMessages.ERROR_CODE_ERROR_RESOLVING_MAIN_APPLICATION.getMessage(),
+                    applicationUuid, tenantDomain);
+            throw new NotificationTemplateManagerException(code, message, e);
+        }
+        validateTemplateLocale(locale);
```

Comment:
Move code duplicaitons to a shared place is possible.

Reply:
addressed with https://github.com/wso2-extensions/identity-event-handler-notification/pull/250/commits/4a4dae818d530081e8a91a33d6d70da13c375c9c

Reply:
This is not complete code duplication and only the logic. Hence will be done seperately after having an internal discussion.